hoodie-client/src/main/java/com/uber/hoodie/index/hbase/HBaseIndex.java

```diff
@@ -370,22 +411,38 @@ private void setPutBatchSize(JavaRDD<WriteStatus> writeStatusRDD,
         If a writeStatus has any insert, it implies that the corresponding task contacts HBase for
         doing puts, since we only do puts for inserts from HBaseIndex.
        */
-      int hbasePutAccessParallelism = getHBasePutAccessParallelism(writeStatusRDD);
+      final Tuple2<Long, Integer> numPutsParallelismTuple  = getHBasePutAccessParallelism(writeStatusRDD);
+      final long numPuts = numPutsParallelismTuple._1;
+      final int hbasePutsParallelism = numPutsParallelismTuple._2;
+      this.numRegionServersForTable = getNumRegionServersAliveForTable();
+      final float desiredQPSFraction = hBaseIndexQPSResourceAllocator.getQPSFractionForPutsTime(numPuts,
+          this.numRegionServersForTable);
+      logger.info("Desired QPSFraction :" + desiredQPSFraction);
+      logger.info("Number HBase puts :" + numPuts);
+      logger.info("Hbase Puts Parallelism :" + hbasePutsParallelism);
+      final float availableQpsFraction = hBaseIndexQPSResourceAllocator.getQPSFraction(desiredQPSFraction, numPuts);
+      logger.info("Allocated QPS Fraction :" + availableQpsFraction);
       multiPutBatchSize = putBatchSizeCalculator
           .getBatchSize(
-              getNumRegionServersAliveForTable(),
+              numRegionServersForTable,
               maxQpsPerRegionServer,
-              hbasePutAccessParallelism,
+              hbasePutsParallelism,
               maxExecutors,
               SLEEP_TIME_MILLISECONDS,
-              qpsFraction);
+              availableQpsFraction);
+      logger.info("multiPutBatchSize :" + multiPutBatchSize);
     }
   }
 
   @VisibleForTesting
-  public int getHBasePutAccessParallelism(final JavaRDD<WriteStatus> writeStatusRDD) {
-    return Math.toIntExact(Math.max(writeStatusRDD
-        .filter(w -> w.getStat().getNumInserts() > 0).count(), 1));
+  public Tuple2<Long, Integer> getHBasePutAccessParallelism(final JavaRDD<WriteStatus> writeStatusRDD) {
+    final JavaRDD<WriteStatus> filteredWriteStatusRDD = writeStatusRDD.filter(w -> w.getStat().getNumInserts() > 0);
+    final JavaPairRDD<Long, Integer> putsParallelismRDD = filteredWriteStatusRDD
+                                                        .mapToPair(w -> new Tuple2<>(w.getStat().getNumInserts(), 1));
+
+    final Tuple2<Long, Integer> putsAndParallelismTuple = putsParallelismRDD
```

Comment:
Can you explain what is the intention of adding the contents of `c` when reducing ?

Reply:
I want to calculate both 1) the number of inserts as well as 2)the number of WriteStatus that has inserts, while reducing.
I could do a count on `writeStatusRDD.filter(w -> w.getStat().getNumInserts() > 0)`  but since I want to reduce the rdd anyway, I am counting there. Let me know if I am missing anything.

Reply:
okay, makes sense.